src/hooks/neo4j.js

```diff
@@ -0,0 +1,73 @@
+import { safeRequire } from '../utils/requireUtils';
+import * as logger from '../logger';
+import { hook } from '../extender';
+import { getRandomId } from '../utils';
+import { SpansContainer, TracerGlobals } from '../globals';
+import { extendNeo4jSpan, createNeo4jSpan } from '../spans/neo4jSpan';
+import { payloadStringify } from '../utils/payloadStringify';
+import { getCurrentTransactionId, NEO4J_SPAN } from '../spans/awsSpan';
+
+function queryBeforeHook(args, extenderContext) {
+  const awsRequestId = TracerGlobals.getHandlerInputs().context.awsRequestId;
+  const transactionId = getCurrentTransactionId();
+  const query = args[0];
+  const params = args[1];
+  const spanId = getRandomId();
+  const started = Date.now();
+
+  const connectionHolder = this._connectionHolderWithMode(this._mode);
+  const connectionParameters = {
+    mode: this._mode,
+    host: connectionHolder._connectionProvider._seedRouter._host,
+    port: connectionHolder._connectionProvider._seedRouter._port,
+    database: connectionHolder._database,
+    user: connectionHolder._connectionProvider._authToken.principal,
+  };
+
+  const span = createNeo4jSpan(
+    transactionId,
+    awsRequestId,
+    spanId,
+    { started },
+    { connectionParameters, query, params },
+    NEO4J_SPAN
+  );
+
+  SpansContainer.addSpan(span);
+  extenderContext.currentSpan = span;
+}
+
+const createResultHook = currentSpan => originalResult => {
+  const ended = Date.now();
+  const extendedSpan = extendNeo4jSpan(currentSpan, {
+    ended,
+    response: payloadStringify(originalResult.records),
+    summary: payloadStringify(originalResult.summary),
+  });
+  SpansContainer.addSpan(extendedSpan);
+};
+
+const createErrorHook = currentSpan => error => {
+  const ended = Date.now();
+  const extendedSpan = extendNeo4jSpan(currentSpan, {
+    ended,
+    error: payloadStringify(error),
+  });
+  SpansContainer.addSpan(extendedSpan);
+};
+
+function queryAfterHook(args, originalFnResult, extenderContext) {
+  const { currentSpan } = extenderContext;
+  originalFnResult.then(createResultHook(currentSpan)).catch(createErrorHook(currentSpan));
```

Comment:
Just note that In this way, you adding the user `catch` handler to the promise also if he didn't ask for it.
It can change the user behavior, We didn't figure out yet what is the best way to hook promises, with the minimum side effect so currently we used `hookPromise` that effect only if the user adds  `then` and `catch` by himself, but for now, you can keep this behavior only for `neo4j` until we will decide what is the best way to instrument promises

Reply:
@doriaviram Ho, that is a good point! Didn't think about that.
Just thinking out loud here, but wouldn't it work to just re-throw the Error after catching it?

Here is a quick runkit demo: https://runkit.com/bboure/wrap-promises

or am I missing something?

Reply:
Yes, it's a good point, actually, You can use that, please add tests for all the cases you can think of, see that the exception is really thrown for example

Reply:
It also is better implantation of our `hookPromise` in the `extender` lib

Reply:
@doriaviram After some investigation and testing, it turns out that this should be working as is.

The reason for that is that we return [the original promise](https://github.com/lumigo-io/lumigo-node/blob/master/src/extender.js#L21) to the user. The `then().catch()` inside the hook has **no effect** at all on it. The original will still throw an error if `then()` (or await) is used without `catch`.
This would not be the case if we had chained them, though.

See demo: https://runkit.com/bboure/chained-promise-vs-non-chained

I also tried using `hookPromise()` first. It turns out that it does not (always) work.
When I used the `await` style promise handling, the span would not be captured.

Reply:
Cool, I will run our integration-test and will keep you updated